feat: add provenance.jq

mrjoelkamp opened the pull request:
## summary
- adds `provenance.jq` filter to generate in-toto provenance for GHA builds

## usage
example usage:
```bash
json="$(
  jq -L.scripts '
    include "meta";
    .[env.BUILD_ID]
    | select(needs_build and .build.arch == env.BASHBREW_ARCH) # sanity check
    | .commands = commands
  ' builds.json
)"

image-digest() {
  local dir="$1"
  local manifest
  manifest="$dir/blobs/$(jq -r '.manifests[0].digest | sub(":"; "/")' "$dir/index.json")" || return "$?"
  jq -L.scripts -r -s '
    include "oci";
    . | image_digest("linux"; env.BASHBREW_ARCH)
  ' $manifest || return "$?"
}
digest=$(image-digest temp)

echo $buildJson | jq -L.scripts --argjson github '${{ env.GITHUB_CONTEXT }}' --argjson runner '${{ toJson(runner) }}' --arg digest ${digest} '
  include "provenance";
  github_actions_provenance($github; $runner; $digest)
' >> provenance.json
```

## output
example output:
```json
{
  "_type": "https://in-toto.io/Statement/v1",
  "subject": [
    {
      "name": "pkg:docker/notary:server-0.7.0?platform=linux%2Famd64",
      "digest": {
        "sha256": "b58c16c79286924b4b96f11061c34498fe137db319aa3e5b606e309f492f2edb"
      }
    },
    {
      "name": "pkg:docker/notary:server?platform=linux%2Famd64",
      "digest": {
        "sha256": "b58c16c79286924b4b96f11061c34498fe137db319aa3e5b606e309f492f2edb"
      }
    },
    {
      "name": "pkg:docker/amd64/notary:server-0.7.0?platform=linux%2Famd64",
      "digest": {
        "sha256": "b58c16c79286924b4b96f11061c34498fe137db319aa3e5b606e309f492f2edb"
      }
    },
    {
      "name": "pkg:docker/amd64/notary:server?platform=linux%2Famd64",
      "digest": {
        "sha256": "b58c16c79286924b4b96f11061c34498fe137db319aa3e5b606e309f492f2edb"
      }
    },
    {
      "name": "pkg:docker/oisupport/staging-amd64:96b43f70bdd4601f5dd42b5525f9818e4b834fde87a0311e28401b4c9a836ff4?platform=linux%2Famd64",
      "digest": {
        "sha256": "b58c16c79286924b4b96f11061c34498fe137db319aa3e5b606e309f492f2edb"
      }
    }
  ],
  "predicateType": "https://slsa.dev/provenance/v1",
  "predicate": {
    "buildDefinition": {
      "buildType": "https://actions.github.io/buildtypes/workflow/v1",
      "externalParameters": {
        "inputs": {
          "bashbrewArch": "amd64",
          "buildId": "96b43f70bdd4601f5dd42b5525f9818e4b834fde87a0311e28401b4c9a836ff4",
          "firstTag": "notary:server-0.7.0"
        },
        "workflow": {
          "ref": "refs/heads/main",
          "repository": "https://github.com/mrjoelkamp/gha-test",
          "path": ".github/workflows/test.yml",
          "digest": {
            "sha256": "11d1f3b6c0efade1b907d0ff3ab41e960787fbf4"
          }
        }
      },
      "internalParameters": {
        "github": {
          "event_name": "workflow_dispatch",
          "repository_id": "712982891",
          "repository_owner_id": "2976326"
        }
      },
      "resolvedDependencies": [
        {
          "uri": "git+https://github.com/mrjoelkamp/gha-test@refs/heads/main",
          "digest": {
            "gitCommit": "11d1f3b6c0efade1b907d0ff3ab41e960787fbf4"
          }
        }
      ]
    },
    "runDetails": {
      "builder": {
        "id": "https://github.com/mrjoelkamp/gha-test/.github/workflows/test.yml@refs/heads/main"
      },
      "metadata": {
        "invocationId": "https://github.com/mrjoelkamp/gha-test/actions/runs/10780651714/attempts/1"
      }
    }
  }
}
```

tianon commented:
I think the shape of this is probably roughly OK -- I've got a lot of minor comments, but I don't want the higher-level discussion to get lost in them, so I'll save them for now. :smile:

I am concerned by how GitHub-centric it is.  Perhaps slightly more accurately, I'm concerned that there aren't any obvious indications or guardrails or places to put a clear TODO for non-GitHub usage of code that is mostly generic. :eyes:

Do you have any thoughts on how/where this might change if we were to generate it from Jenkins also?  Which values would need to change?  To be clear, I don't think we need to do _all_ the work for that, but IMO it's worth annotating at least the things we know are GitHub-only and/or somewhere in the file adding a note that makes it really clear this is GitHub specific for now, perhaps even naming the function `github_actions_provenance` or something instead?  (Thinking through it more, that last one is probably enough to make me feel satisfied :joy:)

Do you want me to go through nits here/now, or wait until we're closer to integrating before hitting the low-level/cosmetic stuff?

whalelines commented on provenance.jq:
```diff
@@ -0,0 +1,78 @@
+def tags:
+  .source.arches[].tags[],
+  .source.arches[].archTags[],
+  .build.img
+;
```
What does this return? Should this function be commented like the others?

mrjoelkamp replied:
an array of image tags, updated comment, thanks 👍

whalelines replied:
Does it need square brackets around it?

```console
$ echo '{"a":[0,1,2,3,4],"b":["foo","bar","baz"],"c":"see"}' | jq '.a[], .b[], .c'
0
1
2
3
4
"foo"
"bar"
"baz"
"see"
```

mrjoelkamp replied:
Collapsing it back down would be reverting some recent changes, see https://github.com/docker-library/meta-scripts/pull/76#discussion_r1763228481

mrjoelkamp replied:
> Does it need square brackets around it?

It doesn't need them but we added them for completeness

whalelines reviewed (commented):
One question

tianon commented on provenance.jq:
```diff
@@ -0,0 +1,80 @@
+        runDetails: {
+            builder: {
+                id: ($context.server_url+"/"+$context.workflow_ref),
+            },
```
There's probably a more interesting data we could include here eventually, like `version` https://slsa.dev/spec/v1.0/provenance#buildmetadata :eyes:

More generally, there's probably even more data in our "build" object that would be interesting to inject here, but I'm not sure exactly where it fits without just blazing our own trail into "extensions" territory (maybe that _is_ the intended way to do so?)

mrjoelkamp replied:
Yes, we could add more at some point. For now, I stuck with the minimum required for the [build type](https://github.com/slsa-framework/github-actions-buildtypes/tree/main/workflow/v1) and that also allows us to form expectations for verification of the build system. 

We will be verifying the subject, external/interal parameters and resolved dependencies using the `builder.id` to bind to a public key and overall build level.

tianon reviewed (commented):
A few more specific comments to hopefully generate more discussion to fine-tune a few points before getting into nits like whitespace :see_no_evil: :heart:

mrjoelkamp commented on provenance.jq:
```diff
@@ -0,0 +1,80 @@
+}
+;
+
+# input: "tags" object with platform, image digest and GITHUB context arguments
+# output: json object for in-toto provenance statement
+def github_actions_provenance($platform; $digest; $context):
```
It was originally collapsed down and everything took `builds.json` (see https://github.com/docker-library/meta-scripts/commit/6fcc5db4fa9e386d169afb28b8c3d714a372741f) but we split it apart so that the build job would output `platform` and `tags` such that the `sign` job wouldn't need to checkout the large `builds.json` because it only needs build tags and platform data. 

We could switch this back but would want @LaurentGoderre to weigh in too since we made this change together.

mrjoelkamp replied:
> maybe there's a subset of that full context document we could pass in/around?

This is a good point, we could just pass in all of the individual fields from the `github` context that we need for provenance. We aren't using any sensitive fields for this but they do exist in the `github` context. Specifically, `github.token`. I am not very worried about the usage here but if you want to we can selectively pass in what we need, maybe make a JSON object from the fields (but not using JQ?) and pass it in that way.

mrjoelkamp replied:
here is a list of what we are currently using:
```
server_url
repository
repository_id
repository_owner_id
ref
workflow_ref
workflow_sha
event_name
event.inputs
run_id
run_attempt
```

mrjoelkamp replied:
so, alternatively we could set the context var like so:
```bash
context=$(jq -n \
  --arg server_url "${{ github.server_url }}" \
  --arg repository "${{ github.repository }}" \
  --arg repository_id "${{ github.repository_id }}" \
  --arg repository_owner_id "${{ github.repository_owner_id }}" \
  --arg ref "${{ github.ref }}" \
  --arg workflow_ref "${{ github.workflow_ref }}" \
  --arg workflow_sha "${{ github.workflow_sha }}" \
  --arg event_name "${{ github.event_name }}" \
  --arg event_inputs "${{ github.event_inputs }}" \
  --arg run_id "${{ github.run_id }}" \
  --arg run_attempt "${{ github.run_attempt }}" \
  '{
    server_url: $server_url,
    repository: $repository,
    repository_id: $repository_id,
    repository_owner_id: $repository_owner_id,
    ref: $ref,
    workflow_ref: $workflow_ref,
    workflow_sha: $workflow_sha,
    event_name: $event_name,
    event_inputs: $event_inputs,
    run_id: $run_id,
    run_attempt: $run_attempt
  }')
```

It may be worth it to make sure the token isn't mishandled. Thoughts on adding that?

mrjoelkamp replied:
I prefer the simplicity of just passing the entire `github` context as a `--jsonarg` since I think the risk of mishandling the `github.token` is relatively low (especially since it is masked in the output). But willing to go either way or see if you have any other ideas for it.

tianon replied:
> It was originally collapsed down and everything took `builds.json` (see [6fcc5db](https://github.com/docker-library/meta-scripts/commit/6fcc5db4fa9e386d169afb28b8c3d714a372741f)) but we split it apart so that the build job would output `platform` and `tags` such that the `sign` job wouldn't need to checkout the large `builds.json` because it only needs build tags and platform data from that file (and we are already consuming it in the build job).

Hmm yeah, I see your point, although that does also make it slightly harder to add more interesting information from it to this document later, if we wanted to. :thinking:

Do you think there would be much overhead in doing something like providing the full "build" object as an output (in JSON format) so that we could pass along _just_ the object from the large file that we're working on?  That would give us something somewhere in the middle so that this is still consistent with most of our other interfaces, easier to run/test locally, and still avoiding the full clone in the provenance generation step. :eyes:

> I prefer the simplicity of just passing the entire `github` context as a `--jsonarg` since I think the risk of mishandling the `github.token` is relatively low (especially since it is masked in the output). But willing to go either way or see if you have any other ideas for it.

Yeah, it's really not ideal either way -- I with GitHub would add more functionality for filtering/munging objects in their expressions so it didn't have to be _so_ manual (or provide a second copy of the context/a function that filters out any sensitive data and makes it safer for raw consumption/use). :confused:

So I guess passing the full context is probably the lesser of two evils and we'll continue to be careful about what we do with it -- we should probably also include a warning similar to that in the upstream documentation to help remind us that the context object is generally sensitive when used in full. :eyes:

Also, as a general point, I think we should avoid using `${{ ... }}` entirely _directly_ in `run:` blocks to help practice our discipline of separation and avoid accidents (https://docs.github.com/en/actions/security-for-github-actions/security-guides/security-hardening-for-github-actions#using-an-intermediate-environment-variable), so that should go in an environment variable, which can then be used in `jq` directly via something like `(env.GITHUB_CONTEXT | fromjson) as $github | ...`, which _also_ helps avoid command-line length limits (and instead pushes them into environment size limits, but those limits are much higher). :smile:

mrjoelkamp replied:
Using environment variable works well, fixed that. We are working on refactoring/testing the build outputs now

mrjoelkamp commented on provenance.jq:
```diff
@@ -0,0 +1,80 @@
+  .build.img
+;
+
+# input: "build" object (with "buildId" top level key)
+# output: purl platform query string
+def platform_string:
+  .source.arches[].platformString | gsub("/"; "%2F")
+;
```
that would be good! If so, we can remove this entirely from here. @LaurentGoderre, lets see if we can use the `uri` filter for the platform output. We just need to make sure that the substitution is there since PURL uses `%2F` not `/` in the string.

mrjoelkamp replied:
@LaurentGoderre was able to use `@uri` outside of `provenance.jq` for `platform_string` so I removed this function entirely

mrjoelkamp commented on provenance.jq:
```diff
@@ -0,0 +1,80 @@
+# output: json object for in-toto provenance subject field
+def subjects($platform; $digest):
+  {
+      "name": ("pkg:docker/" + . + "?platform=" + $platform),
```
I followed existing provenance statements generated by buildkit and used `pkg:docker` to match them.

For the provenance `subject` these items are [ResourceDescriptors](https://github.com/in-toto/attestation/blob/main/spec/v1/resource_descriptor.md) that define digest as a field. So either way, I think both would work.

mrjoelkamp commented on provenance.jq:
```diff
@@ -0,0 +1,80 @@
+  {
+      "name": ("pkg:docker/" + . + "?platform=" + $platform),
+      "digest": {
+        "sha256": $digest
```
For the ResourceDescriptor type the scheme is separate https://github.com/in-toto/attestation/blob/main/spec/v1/digest_set.md

It is being split using parameter expansion `--arg digest ${digest#sha256:}`

mrjoelkamp replied:
I changed this to use `split()` and I do like this better than using the parameter expansion and hard coding the algorithm 👍 

```jq
def subjects($platform; $digest):
  ($digest | split(":")) as $splitDigest |
  {
      "name": ("pkg:docker/" + . + "?platform=" + $platform),
      "digest": {
        "sha256": $digest
        ($splitDigest[0]): $splitDigest[1]
      }
  }
;
```

mrjoelkamp commented on provenance.jq:
```diff
@@ -0,0 +1,80 @@
+                id: ($context.server_url+"/"+$context.workflow_ref),
+            },
+            metadata: {
+                invocationId: ($context.server_url+"/"+$context.repository+"/actions/runs/"+$context.run_id+"/attempts/"+$context.run_attempt),
```
yeah it is odd, but I think it really is the only way to get it

mrjoelkamp commented on provenance.jq:
```diff
@@ -0,0 +1,80 @@
+        },
+        runDetails: {
+            builder: {
+                id: ($context.server_url+"/"+$context.workflow_ref),
```
Yes, I have been going back and forth on what to do here. One option would be to tag the workflow commit and use the tag in the `workflow_dispatch` such that `$context.workflow_ref` is versioned via tagging. But using tags didn't seem like the best approach.

Another option would be to include the commit here. `($context.server_url+"/"+$context.workflow_ref+"@"+$context.workflow_sha)` but this will effectively make any change to `build.yml` ripple into the verification side as we need to be able to identify the builder and a new `sha` would be a new builder. Maybe that is acceptable, how often do we expect `build.yml` to change?

mrjoelkamp replied:
At this point in time `($context.server_url+"/"+$context.workflow_ref)` will uniquely identify this workflow based on our build evaluation until we make a change to `build.yml` that invalidates that.

mrjoelkamp replied:
After a discussion about this, I believe we decided to leave this as-is currently as the ID here will point to the analysis for the build platform as implemented after changes to meet SLSA build level 3. At any point there is a change the invalidates the security properties of the build platform then we would rev the builder ID accordingly (and rotate the public key bound to it).

tianon replied:
Yeah, that seems OK -- I don't love that we might forget / miss a change that _should've_ changed this, but it's fair.  Perhaps we can add a comment here to remind us when we review in the future that this field needs careful review and that _some_ changes mean we should both change it and rotate the key? (especially noting here that if it ever changes, we should also rotate the key :eyes:)

mrjoelkamp replied:
added
```sh
# builder.id identifies the transitive closure of the trusted build platform evalution.
# any changes that alter security properties or build level must update this ID and rotate the signing key.
# https://slsa.dev/spec/v1.0/provenance#builder
```

we'll want to add a comment to the overall `build.yml` as well but will be in a separate PR

mrjoelkamp commented on provenance.jq:
```diff
@@ -0,0 +1,80 @@
+    repository_owner_id: $context.repository_owner_id,
+  }
+}
+;
```
I moved `github_internal_parameters` to be embedded in the main function and kept `github_external_parameters` separate

tianon replied:
Yeah, given they're documents embedded inside well-named keys, I think it's roughly equal (preamble of `def foo:` vs preamble of `foo:`) with a slight preference for embedding because then it matches the upstream documentation / output document slightly better.

tianon commented on provenance.jq:
```diff
@@ -0,0 +1,73 @@
+def subjects($platform; $digest):
+	($digest | split(":")) as $splitDigest |
+	{
+		"name": ("pkg:docker/" + . + "?platform=" + $platform),
```
You needed `@uri` on this platform value, right?

```suggestion
		"name": "pkg:docker/\(.)?platform=\($platform | @uri)",
```

Do we need it on the repo/tag name as well?  Also we need to replace the tag separator with `@`, right? which is kind of hard to do 100% correctly without writing a full "reference parser" :sob:  we can maybe cheat here and assume we'll never have a by-digest reference or a non-standard-port registry and do something like `sub(":"; "@")` but it's a little gross because it's so inaccurate and has several obvious edge cases :disappointed:

mrjoelkamp replied:
we actually have `@uri` here https://github.com/docker-library/meta-scripts/pull/76/files#diff-10e5b6b785f9538d42d6b75a65f23340111a8d9427518b516760afcbb7df75e2R42 where the variable is assigned

the repo/tag name doesn't use the URL encoding AFAICT, we have this working e2e now with signing/verification

mrjoelkamp replied:
fixed the string insertion

tianon commented on provenance.jq:
```diff
@@ -0,0 +1,73 @@
+	{
+		"name": ("pkg:docker/" + . + "?platform=" + $platform),
+		"digest": {
+			($splitDigest[0]): $splitDigest[1]
```
```suggestion
			($splitDigest[0]): $splitDigest[1],
```

tianon commented on provenance.jq:
```diff
@@ -0,0 +1,73 @@
+# input: "tags" object with image digest and platform arguments
+# output: json object for in-toto provenance subject field
+def subjects($platform; $digest):
+	($digest | split(":")) as $splitDigest |
+	{
```
```suggestion
	($digest | split(":")) as $splitDigest
	| {
```

tianon commented on provenance.jq:
```diff
@@ -0,0 +1,76 @@
+# input: GITHUB context argument
+# output: json object for in-toto provenance external parameters field
+def github_external_parameters($github):
+	($github.workflow_ref | gsub( $github.repository + "/"; "")) as $workflowPathRef |
+	{
```
`gsub` does a regex match, so I think we should probably use `ltrimstr` instead to trim the prefix instead, and probably `split` up here too (so we don't do that twice) -- it's also wild, yet again, that `workflow_ref` is the best granularity they offer here and it's this weird string that's unlike anything used anywhere else in the GitHub platform (`octocat/hello-world/.github/workflows/my-workflow.yml@refs/heads/my_branch`) :sob:

```suggestion
	($github.workflow_ref | ltrimstr($github.repository + "/") | split("@")) as $workflowRefSplit
	| {
```

tianon commented on provenance.jq:
```diff
@@ -0,0 +1,76 @@
+			ref: ($workflowPathRef | split("@")[1]),
+			repository: ($github.server_url + "/" + $github.repository),
+			path: ($workflowPathRef | split("@")[0]),
+			digest: {sha256: $github.workflow_sha}
```
`workflow_sha` is just a commit digest, right?  so this should be either `sha1` or better yet `gitCommit` again?

```suggestion
			digest: { gitCommit: $github.workflow_sha },
```

(doesn't this actually belong in the "resolved dependencies" section?)

mrjoelkamp replied:
yes, thank you

> (doesn't this actually belong in the "resolved dependencies" section?)

this was an addition I made to record the commit sha of the workflow (since it wasn't anywhere else) and the other fields here identify the workflow as part of https://github.com/actions/buildtypes/tree/main/workflow/v1

to me it was missing the commit, so I added it. In our use-case it is a bit redundant since the `meta` resolvedDependency == workflow commit

tianon reviewed (commented):
I'm not completely done, and I think there's still some discussion points open (notably the high-level input) so I'm trying to avoid commenting in places that might change or go away based on those discussions, but here's a few more nit-picking notes before I go to lunch :eyes:

mrjoelkamp commented:
> I'm not completely done, and I think there's still some discussion points open (notably the high-level input) so I'm trying to avoid commenting in places that might change or go away based on those discussions, but here's a few more nit-picking notes before I go to lunch 👀

@LaurentGoderre and I hit most of these comments and we also refactored the high level input to use the build object for the `buildId` as an output from the build job pass into the sign job to hopefully resolve https://github.com/docker-library/meta-scripts/pull/76#discussion_r1765457800

Remember that we have a working test environment and some of the high-level changes aren't captured here but will be in a future PR with changes to `build.yml`. all of these changes are tested and verified working in GHA in a separate repo.

tianon commented:
I had more suggestions, but figured they might be easier to digest as explicit commits instead of a pile of interconnected suggestions (or prose / whatever it is that GitHub does with suggestion blocks that span a large number of lines):

- https://github.com/docker-library/meta-scripts/commit/96f29bc091b013eca44d873e65daa0851a1c6fb6

  > Collapse more functions into main doc, add more trailing commas, minor calculation adjustments
  >
  > The primary justification for collapsing everything back into the main document (aside from there only being a single caller for each function) is that it makes verifying that we didn't miss anything easier -- as you scan the document, for each field that contains a "load-bearing" URL (first https://in-toto.io/Statement/v1, then https://slsa.dev/provenance/v1, and finally https://actions.github.io/buildtypes/workflow/v1), if you open the URL it describes the expected format, fields, and values, and with them all here and in-order, they're *much* easier to match up and validate to be correct and exhaustive.  Granted, that will change over time as we shove more and more (optional) data into this document so that it includes a more complete picture, but for now, this is makes it really easy to double check our work (and the end result is no less organized; for example, the `externalParameters` are still all grouped together under a suitable heading describing what their purpose is).
  >
  > I also made some minor changes to the way values were calculated, especially in the `workflow` block, but very related to the above justification: now the way we calculate the values matches the way they're described in https://actions.github.io/buildtypes/workflow/v1 (specifically using the exact fields parsed in the exact ways they suggest).  We will probably deviate from that over time (as suggested by a new "TODO" comment I included), but at least this way our baseline matches theirs and the delta will be easier to track.
  >
  > Additionally, I removed the `(env.GITHUB_CONTEXT | fromjson) as $github` line from here, because I think that's more appropriate behavior for the caller (and added back the explicit function arguments).  This will be more clearly meaningful in my follow-up commit adding a basic test.

- https://github.com/docker-library/meta-scripts/commit/ad28b37e88258c159f9480063ef8165b4ee40973

  > Add a basic test of the provenance (both amd64 and windows, to illustrate/validate the `inputs` variance)

You can also view them via https://github.com/docker-library/meta-scripts/compare/24e679e89987a94f28285b27eb8ba4dab5c3a224...ad28b37e88258c159f9480063ef8165b4ee40973 or steal them directly via fetch-by-commit or https://github.com/infosiftr/doi-meta-scripts/tree/tianon-provenance if you find them unobjectionable as-is.  (Of course, happy to discuss further, explain more/better, push directly, adjust, etc etc.)

mrjoelkamp commented:
> You can also view them via [24e679e...ad28b37](https://github.com/docker-library/meta-scripts/compare/24e679e89987a94f28285b27eb8ba4dab5c3a224...ad28b37e88258c159f9480063ef8165b4ee40973) or steal them directly via fetch-by-commit or https://github.com/infosiftr/doi-meta-scripts/tree/tianon-provenance if you find them unobjectionable as-is. (Of course, happy to discuss further, explain more/better, push directly, adjust, etc etc.)

@tianon 🙇 thanks for making the time to take a pass on this!  I pushed your commits since I don't really have any objections and like the way it turned out.

I also added the `runner` context, in order to not hardcode `runner.environment` in [d9b5735](https://github.com/docker-library/meta-scripts/pull/76/commits/d9b57350b26c4d28a73e93f2aa64495c9366019c). 

I didn't realize there was a whole test framework here 😆 , appreciate the addition and updated it with the `runner` context.

mrjoelkamp commented:
>   > I also made some minor changes to the way values were calculated, especially in the `workflow` block, but very related to the above justification: now the way we calculate the values matches the way they're described in https://actions.github.io/buildtypes/workflow/v1 (specifically using the exact fields parsed in the exact ways they suggest).  We will probably deviate from that over time (as suggested by a new "TODO" comment I included), but at least this way our baseline matches theirs and the delta will be easier to track.

I might want to revert the workflow back to what we had before. Specifically, because this really should represent the workflow claims. As implemented in https://github.com/actions/attest-build-provenance for example, they also decided to use the `github.workflow_ref` as a basis for the workflow metadata: 
```js
    const [workflowPath, workflowRef] = claims.workflow_ref
        .replace(`${claims.repository}/`, '')
        .split('@');
```
...
```js
                externalParameters: {
                    workflow: {
                        ref: workflowRef,
                        repository: `${serverURL}/${claims.repository}`,
                        path: workflowPath
                    }
                },
```

I believe this is more accurate despite the example in the [build type](https://github.com/actions/buildtypes/tree/main/workflow/v1) definition.

tianon commented:
The problem is we don't have a `workflow_repository` variable, so we're still relying on the context being the same so it didn't feel worthwhile IMO to only fix that dissonance halfway :thinking:

tianon commented on provenance.jq:
```diff
@@ -0,0 +1,75 @@
+						id: ($github.server_url + "/" + $github.workflow_ref),
+					},
+					metadata: {
+						invocationId: "\($github.server_url)/\($github.repository)/actions/runs/\($github.run_id)/attempts/\($github.run_attempt)",
```
I joined this because I had (erroneously) assumed `run_id` and `run_attempt` would be numeric and run into data joining issues, but they're both strings, so this can be reverted to the `+` version that more obviously matches the upstream version:

```suggestion
						invocationId: ($github.server_url + "/" + $github.repository + "/actions/runs/" + $github.run_id + "/attempts/" + $github.run_attempt)
```

tianon commented on provenance.jq:
```diff
@@ -0,0 +1,75 @@
+# output: in-toto provenance statement (https://slsa.dev/spec/v1.0/provenance)
+#   see also: https://github.com/actions/buildtypes/tree/main/workflow/v1
+def github_actions_provenance($github; $runner; $digest):
+	if $github.event_name != "workflow_dispatch" then error("error: '\($github.event_name)' is not a supported event type for provenance generation") else
```
Oh, I forgot to mention this in my commit message, so I'll highlight it with a comment instead -- I was looking at https://github.com/actions/buildtypes/tree/main/workflow/v1 and the list of events it supports, and realized our provenance has even tighter assumptions than theirs (especially for `$github.event.inputs`) so made anything else an error accordingly just to be safe. :+1:

mrjoelkamp replied:
yes, I noticed this and definitely like the addition 👍

tianon commented on provenance.jq:
```diff
@@ -0,0 +1,75 @@
+					},
+					resolvedDependencies: [
+						{
+							uri: "git+\($github.server_url)/\($github.repository)@\($github.ref)",
```
This one could also be de-interpolated if you prefer -- I don't have a strong preference either way.

tianon commented on provenance.jq:
```diff
@@ -0,0 +1,75 @@
+#   see also: https://github.com/actions/buildtypes/tree/main/workflow/v1
+def github_actions_provenance($github; $runner; $digest):
+	if $github.event_name != "workflow_dispatch" then error("error: '\($github.event_name)' is not a supported event type for provenance generation") else
+	($github.workflow_ref | ltrimstr($github.repository + "/") | split("@") | if length == 2 then . else error("parsing 'workflow_ref' failed: '\(.)'") end) as $workflowRefSplit
```
If we stick with this or something like it (https://github.com/docker-library/meta-scripts/pull/76#issuecomment-2361475268), it should move down into the workflow block, something like this:

```jq
	workflow: (
		($github.workflow_ref | ...) as $workflowRefSplit # or $workflowRefParts? if we split it more 🤔  with the capturing regex approach I put in my TODO comment, it could be the basis for the object and not `as` at all, instead just `+ { the missing bits }` 👀
		| {
			...
		}
	),
```

tianon reviewed (approved):
:eyes:
